### PR TITLE
1463 mean absolute percentage error

### LINKED ISSUE #1463: Add Mean Absolute Percentage Error (MAPE)
For timeseries.

https://en.wikipedia.org/wiki/Mean_absolute_percentage_error

@freddyaboulton

### PR DESCRIPTION
**Pull request by ParthivNaresh**
Fixes #1463

### CODE REVIEW

File: evalml/objectives/standard_metrics.py
Comment (freddyaboulton) [inline evalml/objectives/standard_metrics.py:355]: Maybe we should just raise a ValueError if any `y_true` are 0? My concern is that 0 can be a meaningful target value (like stock returns) and users may be mislead that those records are being used in the evaluation when they are not.

File: evalml/objectives/standard_metrics.py
Comment (freddyaboulton) [inline evalml/objectives/standard_metrics.py:352]: nit-pick: You don't need to specify `score_needs_proba` since that is specified in `RegressionObjective`

File: evalml/objectives/standard_metrics.py
Comment (freddyaboulton) [inline evalml/objectives/standard_metrics.py:348]: I think we should add to the api reference too

**Review by freddyaboulton — APPROVED** [review]
@ParthivNaresh This looks great! I think we should resolve the discussion about what to do with division by 0 before merging but other than that I think it's good to merge! 

I'm ok with holding off with making this the default objective for time series regression problems because of its [noted limitations](https://en.wikipedia.org/wiki/Mean_absolute_percentage_error#Issues), namely the bias in selecting estimators that under-estimate the target.

I think for the default objective for automl we should consider these two alternatives:
1. [Mean Absolute Scaled Error](https://en.wikipedia.org/wiki/Mean_absolute_scaled_error)
2. [Log forecast to Actual](https://poseidon01.ssrn.com/delivery.php?ID=919102102125114084027127082015014024125011038043027056100114112024102116028092091119017122037004103000001023067093098125105096001014042028065078087113084005126022037011055081071017115026000087023008123083066070022120086113103100120020126116115009101&EXT=pdf)

File: evalml/objectives/standard_metrics.py
Comment (dsherry) [inline evalml/objectives/standard_metrics.py:356]: @ParthivNaresh nice, looks good!

Sklearn 0.24.2 (a current release candidate) [has a MAPE impl](https://github.com/scikit-learn/scikit-learn/blob/4773f3e39/sklearn/metrics/_regression.py#L197) ([docs](https://scikit-learn.org/dev/modules/generated/sklearn.metrics.mean_absolute_percentage_error.html)).

I like [their use of `max(y_true, eps)`](https://github.com/scikit-learn/scikit-learn/blob/4773f3e39/sklearn/metrics/_regression.py#L261) in the denominator. I think we should do that instead of excluding rows where `y_true` is exactly 0.

I also liked that they had weight support. We don't have a pattern for this yet from other objectives, and its not a requirement for this PR, but would be nice to have.

File: evalml/objectives/standard_metrics.py
Comment (dsherry) [inline evalml/objectives/standard_metrics.py:349]: Should we mention that we've scaled this by 100 so that its a percentage? I've seen it defined both with or without the 100 factor in the past.

File: evalml/tests/objective_tests/test_standard_metrics.py
Comment (dsherry) [inline evalml/tests/objective_tests/test_standard_metrics.py:455]: @ParthivNaresh nice 👍  I was able to walk through by hand and reproduce the math for the first example you have here. Only thought here is that if you could simplify the numbers a bit, make the arrays shorter,  make some of those decimals more round etc., it would help make this math easier to follow.

File: evalml/objectives/standard_metrics.py
Comment (dsherry) [inline evalml/objectives/standard_metrics.py:348]: @freddyaboulton did we decide not to define `TimeseriesRegressionObjective`? Or did we simply not decide yet, haha, since this is the first objective explicitly defined for timeseries? I can't remember where we left that conversation.

**Review by dsherry — APPROVED** [review]
@ParthivNaresh looks great!! The only blocking comments I had were about using `eps` in the impl, and starting a thread with @freddyaboulton about whether or not we should define `TimeseriesRegressionObjective` right now

File: evalml/objectives/standard_metrics.py
Comment (ParthivNaresh) [reply evalml/objectives/standard_metrics.py:356]: @dsherry That's an interesting approach, but wouldn't that result in an extremely high value in the overall calculation? @freddyaboulton

Comment (freddyaboulton) [reply evalml/objectives/standard_metrics.py:356]: Good point @dsherry! sklearn's definition of `eps` is way too small but even if we used our own I'm worried that there'll always be one use case where that value is "wrong".

I prefer treating this how we treat RootMeanSquaredLogError or MeanSquaredLogError. Those objectives error out if passed targets with any negative values. They are also kept out of the core objectives because the user only uses them when they are defined for their particular dataset. Maybe we should add MAPE to `get_non_core_objectives`

image


image

File: evalml/objectives/standard_metrics.py
Comment (freddyaboulton) [reply evalml/objectives/standard_metrics.py:348]: Previously, our objectives were only defined for a single problem type. We agreed to change this to a list of valid problem types and that our existing regression objectives should work for both regression and time series regression problems.

So long story short, we didn't define a `TimeSeriesRegressionObjective` because we didn't have any objectives that should only work for time series. We should make MAPE a `TimeSeriesRegressionObjective` in that case :) Good call.

Comment (dsherry) [reply evalml/objectives/standard_metrics.py:348]: Ah got it, thanks @freddyaboulton 

So @ParthivNaresh , I think what we should do here is:
* Define `TimeseriesRegressionObjective` which subclasses `RegressionObjective` but allows only `ProblemTypes.TIME_SERIES_REGRESSION`
* Have `MAPE` subclass `TimeseriesRegressionObjective`

That sound good?

File: evalml/objectives/standard_metrics.py
Comment (dsherry) [reply evalml/objectives/standard_metrics.py:356]: > sklearn's definition of eps is way too small

Hmm, its an arbitrary threshold, right? Setting it to min numerical precision makes sense to me because any differences smaller than that (more or less) cannot be trusted.

It seems we have a few options for what to do when the target contains values equal to or super close to zero (options not all mutually exclusive):
1. Have the objective error out.
2. Have the objective round the zero value up to a small eps in the denominator.
3. Have automl error out during init if MAPE was selected as the automl objective. Similar treatment for RMSLE/MSLE if the target has negative values.

What behavior is most useful to users?

@freddyaboulton @ParthivNaresh I don't have a strong preference between 1 and 2--if you're both advocating for 1, let's do that in this PR. However to really nail that question I think we should also file 3 as an enhancement and address it separately.

Comment (freddyaboulton) [reply evalml/objectives/standard_metrics.py:356]: I think the problem with the sklearn approach is that it will divide by eps whenever y=0 which will cause the score to blow up.

image

I don't think scores that large are useful to users, especially if they have a dataset where y=0 occurs frequently. I like option 1 because I prefer letting the user know at the beginning that this objective isn't great for problems with y=0 (rather than having them comb through or impl trying to reason why their objective is ~1e14). 

You bring up a good point about AutoMLSearch! We should definite file that. Currently we never allow RMSLE and MSLE as the primary objective but I like the idea of only disallowing it if the target has negative values. Similar for MAPE. That sounds like it could be a data check!

Comment (dsherry) [reply evalml/objectives/standard_metrics.py:356]: @freddyaboulton agreed! And ooh yeah definitely, perhaps that could be part of the "InvalidTargetsDataCheck"

Comment (freddyaboulton) [reply evalml/objectives/standard_metrics.py:356]: Filed #1521 !

File: evalml/objectives/standard_metrics.py
Comment (ParthivNaresh) [reply evalml/objectives/standard_metrics.py:348]: @dsherry Sounds great!

File: evalml/objectives/standard_metrics.py
Comment (dsherry) [inline evalml/objectives/standard_metrics.py:361]: 👍  cool!

File: evalml/objectives/time_series_regression_objective.py
Comment (dsherry) [inline evalml/objectives/time_series_regression_objective.py:13]: @ParthivNaresh looks great thanks

File: evalml/tests/dependency_update_check/latest_dependency_versions.txt
Comment (dsherry) [inline evalml/tests/dependency_update_check/latest_dependency_versions.txt:8]: Ah @ParthivNaresh for future reference, if the dependency update check fails on your PR, the dependency checker bot which runs every hour will get a separate PR up in with this update. And then once that's merged you can rebase/update your PR and you should be all set.

Comment (dsherry) [reply evalml/tests/dependency_update_check/latest_dependency_versions.txt:8]: We keep those things separate so that we can watch the dependency update step pass/fail CI, independent from any other changes like those being made in this PR. This helps us catch breaking changes made by the latest version of our dependency libs.

**Pull request merged**
